# Microbial ecology of the transplanted human lung

We combined amplicon sequencing and culturomics to characterize the viable bacterial community in 234 longitudinal bronchoalveolar lavage samples from 64 lung transplant recipients and established links to viral loads, host gene expression, lung function, and transplant health. 
We find that the lung microbiota post-transplant can be categorized into four distinct compositional states, ‘pneumotypes’. 

## Code conceptualization and development team 

* Sudip Das<sup>1
* Germán Bonilla-Rosso<sup>1
* Céline Pattaroni<sup>2
* Alexis Rapin<sup>3
* Philipp Engel<sup>1

1 Department of Fundamental Microbiology, Biophore, University of Lausanne, Switzerland.

2 Department of Immunology and Pathology, Central Clinical School, Monash University, Australia.

3 Division of Pneumology, Centre Hospitalier Universitaire Vaudois (CHUV), Lausanne, Switzerland.

## Code environment and structure

**1. System Requirements** 

This pipeline has been tested on the following version of the required software.

* macOS 10.15.4 Catalina or Ubuntu 20.04 or MS Windows 10
* Python 2.7 
* Bash script 
* R version 3.6.3
* [Genocrunch](https://genocrunch.epfl.ch/home/doc) v 2 Release 2018/06/08

**Installation** 

* Python 2.7 comes pre-installed with macOS and Ubuntu. 
* For MS Windows, follow the [installation instructions](https://docs.python.org/3/using/windows.html).
* For MS Windows 10 now comes with Bash Shell scripting option. Intiate this by typing:
```
#!/bin/bash
```
The analysis is divided into 2 parts:


## Data availibity 

We have deposited the raw data from all samples used in the study to Short Read Archive, NCBI under the BioProject **PRJNA632552** and BioSample accession **SAMN14911405**. 


